# Project Description

This project aims to reproduce a portion of the findings found during single cell RNA-seq analysis of pancreatic cells from human and mice samples by Baron et. al (2016). The referenced study used sc-RNA-seq data from 4 human donors and 2 mice strains for single cell analysis. This project is divided into three parts: data curator, programmer, and analyst. It contains the code that was used to pre-process and analyze the down-stream data.

Reference: Baron, M., Veres, A., Wolock, S. L., Faust, A. L., Gaujoux, R., Vetere, A., Ryu, J. H., Wagner, B. K., Shen-Orr, S. S., Klein, A. M., Melton, D. A., & Yanai, I. (2016). A Single-Cell Transcriptomic Map of the Human and Mouse Pancreas Reveals Inter- and Intra-cell Population Structure. Cell systems, 3(4):346â€“360.e4. https://doi.org/10.1016/j.cels.2016.08.011

# Contributors

Data Curator: Teresa Rice - tpillars@bu.edu

Programmer: Arushi Shrivastava - arushi08@bu.edu

Analyst: Maha Naim - mnaim21@bu.edu

# Repository Contents

Files located in our repository are organized according to project role. Explanations as to what each code does and how to execute it are described below. 

# Data Curator
Run files in order listed
## commands1.sh
Updates salmon to salmon 1.4.0, gets transcripts and primary assembly files, generates decoys text for salmon index
## getbarcodes_stat4.py
Reads through gzip file SRR3879604_1_bc.fastq.gz to extract barcodes and collect barcode count
## getbarcodes_stat5.py
Reads through gzip file SRR3879605_1_bc.fastq.gz to extract barcodes and collect barcode count
## getbarcodes_stat6.py
Reads through gzip file SRR3879606_1_bc.fastq.gz to extract barcodes and collect barcode count
## combine_bc.py
Takes output dictionary from all getbarcodes_stat#.py files and combines into one large dictionary to remove duplicates, create .csv
## commands2.sh
Get salmon index, get pre-map file for salmon alevin 
## get_map.py
Reads in pre-map file and completes the formatting, be sure to rename output file to txp2gene.tsv before running salmon alevin
## commands3.sh
Runs the salmon alevin with SRR3879604,SRR3879605,SRR3879606. Takes in txp2gene.tsv for the tgMap and whitelist_bc_thresh100.txt for the whitelist.

# Programmer
### see programmer folder for programmer README

# Analyst
## analyst_code.R 
Contains the R code written to assign cell types to the respective clusters generated by the programmer code. The input file is a Seurat object. Required packages are listed in the file. Heatmaps and feature, violin, and UMAP projection plots are generated depicting marker gene enrichment according to cell type.
